enterprise/cypher/compiled-expressions/src/main/java/org/neo4j/cypher/internal/runtime/compiled/expressions/CypherMath.java

```diff
@@ -0,0 +1,238 @@
+ * GNU Affero General Public License for more details.
+ *
+ * Neo4j object code can be licensed independently from the source
+ * under separate terms from the AGPL. Inquiries can be directed to:
+ * licensing@neo4j.com
+ *
+ * More information is also available at:
+ * https://neo4j.com/licensing/
+ */
+package org.neo4j.cypher.internal.runtime.compiled.expressions;
+
+import org.opencypher.v9_0.util.CypherTypeException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ThreadLocalRandom;
+
+import org.neo4j.internal.kernel.api.CursorFactory;
+import org.neo4j.internal.kernel.api.NodeCursor;
+import org.neo4j.internal.kernel.api.PropertyCursor;
+import org.neo4j.internal.kernel.api.RelationshipScanCursor;
+import org.neo4j.internal.kernel.api.Transaction;
+import org.neo4j.kernel.impl.util.ValueUtils;
+import org.neo4j.values.AnyValue;
+import org.neo4j.values.SequenceValue;
+import org.neo4j.values.ValueMapper;
+import org.neo4j.values.storable.ArrayValue;
+import org.neo4j.values.storable.BooleanValue;
+import org.neo4j.values.storable.DateTimeValue;
+import org.neo4j.values.storable.DateValue;
+import org.neo4j.values.storable.DoubleValue;
+import org.neo4j.values.storable.DurationValue;
+import org.neo4j.values.storable.LocalDateTimeValue;
+import org.neo4j.values.storable.LocalTimeValue;
+import org.neo4j.values.storable.NumberValue;
+import org.neo4j.values.storable.PointValue;
+import org.neo4j.values.storable.StringValue;
+import org.neo4j.values.storable.TemporalValue;
+import org.neo4j.values.storable.TextValue;
+import org.neo4j.values.storable.TimeValue;
+import org.neo4j.values.storable.Value;
+import org.neo4j.values.storable.Values;
+import org.neo4j.values.virtual.ListValue;
+import org.neo4j.values.virtual.MapValue;
+import org.neo4j.values.virtual.PathValue;
+import org.neo4j.values.virtual.VirtualNodeValue;
+import org.neo4j.values.virtual.VirtualRelationshipValue;
+import org.neo4j.values.virtual.VirtualValues;
+
+import static org.neo4j.values.storable.Values.NO_VALUE;
+import static org.neo4j.values.storable.Values.doubleValue;
+import static org.neo4j.values.storable.Values.stringValue;
+
+/**
+ * This class contains static helper math methods used by the compiled expressions
+ */
+@SuppressWarnings( "unused" )
+public final class CypherMath
+{
+    private CypherMath()
+    {
+        throw new UnsupportedOperationException( "Do not instantiate" );
+    }
+
+    //TODO this is horrible spaghetti code, we should push most of this down to AnyValue
+    public static AnyValue add( AnyValue lhs, AnyValue rhs )
+    {
+        if ( lhs == NO_VALUE || rhs == NO_VALUE )
+        {
+            return NO_VALUE;
+        }
+        if ( lhs instanceof NumberValue && rhs instanceof NumberValue )
+        {
+            return ((NumberValue) lhs).plus( (NumberValue) rhs );
+        }
+        //List addition
+        //arrays are same as lists when it comes to addition
```

Comment:
Could we use `SequenceValue` here? It's supposed to unify `ArrayValue` and `ListValue`.

Reply:
the `prepend` and `append` method is only defined on `ListValue`. It should probably be moved to `SequenceValue`. I want to move `add` to live on `AnyValue` so that we have all this in one place. I have added a card for that.